components/weekly_storage/weekly_event_storage.h

```diff
@@ -0,0 +1,136 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_WEEKLY_STORAGE_WEEKLY_EVENT_STORAGE_H_
+#define BRAVE_COMPONENTS_WEEKLY_STORAGE_WEEKLY_EVENT_STORAGE_H_
+
+#include <list>
+#include <memory>
+#include <utility>
+
+#include "base/time/clock.h"
+#include "base/time/default_clock.h"
+#include "base/time/time.h"
+#include "base/values.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+// WeeklyStorage variant holding a list of events over the past week.
+//
+// Mostly used by various P3A recorders to report whether an event happened
+// during the measurement period.
+//
+// New event values are recorded by calling `Add()` and are forgotten
+// after approximately a week.
+//
+// Parametarized over an `enum class` to get some type checking for the caller.
+// Requires |pref_name| to be already registered.
+template <typename T>
```

Comment:
it's generally discouraged to created templated classes in Chromium because of code bloating and binary size increase. Can we just stick with `int` for P3A histogram values?

Reply:
I did it this way so we could have type specificity at the call sites. I don't think type checking here will catch that many mistakes, but it makes for better code completion, and makes it easier to do semantically meaningful reports or cross-checking with the stats back end in the future.

I thought that was worth the bloat. I guess LTO doesn't currently consolidate method implementations for this case?

Reply:
I went ahead an lowered it to `int` values, since @iefremov said something similar.